src/main/java/games/strategy/engine/lobby/server/login/RsaAuthenticator.java

```diff
@@ -0,0 +1,153 @@
+  private static final String RSA = "RSA";
+  private static final String RSA_ECB_OAEPP = RSA + "/ECB/OAEPPadding";
+
+  @VisibleForTesting
+  static final String ENCRYPTED_PASSWORD_KEY = "RSAPWD";
+
+  @VisibleForTesting
+  static final String RSA_PUBLIC_KEY = "RSAPUBLICKEY";
+
+  private static Cache<String, PrivateKey> buildCache() {
+    return CacheBuilder.newBuilder().expireAfterWrite(10, TimeUnit.MINUTES).build();
+  }
+
+  /**
+   * Returns true if the specified map contains the required values.
+   */
+  static boolean canProcessResponse(final Map<String, String> response) {
+    return response.get(ENCRYPTED_PASSWORD_KEY) != null;
+  }
+
+  /**
+   * Returns true if the specified map contains the required values.
+   */
+  public static boolean canProcessChallenge(final Map<String, String> challenge) {
+    return challenge.get(RSA_PUBLIC_KEY) != null;
+  }
+
+  /**
+   * Adds public key of a generated key-pair to the challenge map
+   * and stores the private key in a map.
+   */
+  static void appendPublicKey(final Map<String, String> challenge) {
+    challenge.put(RSA_PUBLIC_KEY, storeKeypair());
+  }
+
+  private static String storeKeypair() {
+    KeyPair keyPair;
+    String publicKey;
+    do {
+      keyPair = generateKeypair();
+      publicKey = Base64.getEncoder().encodeToString(keyPair.getPublic().getEncoded());
+    } while (rsaKeyMap.getIfPresent(publicKey) != null);
+    rsaKeyMap.put(publicKey, keyPair.getPrivate());
+    return publicKey;
+  }
+
+  private static KeyPair generateKeypair() {
+    try {
+      final KeyPairGenerator keyGen = KeyPairGenerator.getInstance(RSA);
+      keyGen.initialize(4096);
+      return keyGen.generateKeyPair();
+    } catch (NoSuchAlgorithmException e) {
+      throw new IllegalStateException(RSA + " is an invalid algorithm!", e);
+    }
+  }
+
+  private static String decryptPassword(final String encryptedPassword, final PrivateKey privateKey,
+      final Function<String, String> successfullDecryptionAction) {
+    Preconditions.checkNotNull(encryptedPassword);
+    Preconditions.checkNotNull(privateKey);
+    try {
+      final Cipher cipher = Cipher.getInstance(RSA_ECB_OAEPP);
+      cipher.init(Cipher.DECRYPT_MODE, privateKey);
+      return successfullDecryptionAction
+          .apply(new String(cipher.doFinal(Base64.getDecoder().decode(encryptedPassword)), StandardCharsets.UTF_8));
+    } catch (NoSuchAlgorithmException | NoSuchPaddingException e) {
+      throw new IllegalStateException(e);
+    } catch (InvalidKeyException | IllegalBlockSizeException | BadPaddingException e) {
+      return e.getMessage();
+    }
+  }
+
+  /**
+   * Attempts to decrypt the given password using the challenge and response parameters.
+   */
+  public static String decryptPasswordForAction(final Map<String, String> challenge, final Map<String, String> response,
+      final Function<String, String> successFullDecryptionAction) {
```

Comment:
Might be good to document what `successFullDecryptionAction` is, what its single parameter is, and what its return value is.  Also, and related, that this method returns an error message, possibly including whatever `successFullDecryptionAction` returns.

Spelling nit: `successFullDecryptionAction` should probably be `successfulDecryptionAction`.